### PR TITLE
[utils] Make distinction between configured and online CPU count

### PR DESCRIPTION
**Pull request by ludovic-henry**

### CODE REVIEW

**Comment by ludovic-henry** [comment]
@akoeplinger could you please review this?

cc @kumpera

**Comment by akoeplinger** [comment]
I'm writing up a more detailed explanation of my findings now (give me a few mins), but I don't think we can change ProcessorCount to return the available CPUs on ARM. The split in general looks good to me though!

**Comment by akoeplinger** [comment]
Ok, so here's the recap:

When we merged the change from PR #2722, we started seeing random failures on ARM in the MonoTests.System.Threading.ThreadPoolTests.SetAndGetMaxThreads and MonoTests.System.Threading.ManualResetEventSlimTests.Constructor_Defaults tests. Both of those tests are dealing with Environment.ProcessorCount to verify some implementation details.

It turns out that on the Jetson TK1 board we use on public Jenkins and on ARM kernels in general, the value returned by sched_getaffinity (or _SC_NPROCESSORS_ONLN) doesn't contain CPUs/cores that are powered off for power saving reasons. This is contrary to what happens on x86, where even cores in deep-sleep state are returned [1], [2]. This means that we would get a processor count of 1 at one point in time and a higher value when load increases later on as the system wakes CPUs.

Various runtime pieces like the threadpool and also user code however relies on the value returned by Environment.ProcessorCount e.g. for deciding how many parallel tasks to start, thereby limiting the performance when that code thinks we only have one CPU.

Talking to a few people, this was the reason why we changed to _SC_NPROCESSORS_CONF in https://github.com/mono/mono/pull/1688 and why we added a special case for Android in https://github.com/mono/mono/commit/de3addcde5bf875dcd43d3e2c6536d61871e7675 to get the "real" number of processors in the system.

Because of those issues Android/Dalvik also switched from _ONLN to _SC_NPROCESSORS_CONF for the Java API Runtime.availableProcessors() too [3], citing:
> Traditionally this returned the number currently online,
but many mobile devices are able to take unused cores offline to
save power, so releases newer than Android 4.2 (Jelly Bean) return the maximum number of
cores that could be made available if there were no power or heat
constraints.

The problem with sticking to _SC_NPROCESSORS_CONF however is that it breaks down in constrained environments like Docker or with an explicit CPU affinity set by the Linux `taskset` command, They'd get a higher CPU count than can be used, start more threads etc. which results in unnecessary context switches and overloaded systems. That's why we need to respect sched_getaffinity.

So while in an ideal world we would be able to rely on sched_getaffinity/_SC_NPROCESSORS_ONLN to return the number of theoretically available CPUs regardless of power saving measures everywhere, we can't do this on ARM.

I think the pragmatic solution is the following:
* use sched_getaffinity (+ fallback to _SC_NPROCESSORS_ONLN in case of error) on x86. This ensures we're inline with what OpenJDK [4] and CoreCLR [5] do
* use _SC_NPROCESSORS_CONF exclusively on ARM (I think we could eventually even get rid of the PLATFORM_ANDROID special case)

Thoughts?

Helpful links:

[1] https://sourceware.org/ml/libc-alpha/2013-07/msg00383.html
[2] https://lists.01.org/pipermail/powertop/2012-September/000433.html
[3] https://android.googlesource.com/platform/libcore/+/750dc634e56c58d1d04f6a138734ac2b772900b5%5E1..750dc634e56c58d1d04f6a138734ac2b772900b5/
[4] https://bugs.openjdk.java.net/browse/JDK-6515172
[5] https://github.com/dotnet/coreclr/blob/7058273693db2555f127ce16e6b0c5b40fb04867/src/pal/src/misc/sysinfo.cpp#L148

File: mono/utils/mono-proclib.c [hunk 1]
```diff
@@ -615,15 +615,18 @@ mono_process_current_pid ()
...
 #ifdef HOST_WIN32
 	SYSTEM_INFO info;
 	GetSystemInfo (&info);
 	return info.dwNumberOfProcessors;
 #else
-	int count = 0;
+
 #ifdef PLATFORM_ANDROID
+	/* You cannot set the processor cpu affinity on android, so there is no difference between
+	 * the number of configured CPU and available CPU, as per a runtime point of view */
```
Comment (akoeplinger) [inline mono/utils/mono-proclib.c]: That's not really true, you could just pinvoke into sched_setaffinity.

File: mono/utils/mono-proclib.c [hunk 2]
```diff
@@ -642,20 +645,38 @@ mono_cpu_count (void)
...
+			count = sysconf (_SC_NPROCESSORS_ONLN);
+#endif
+			break;
+		default:
+			g_assert_not_reached ();
+		}
+		if (count > 0)
+			return count;
+	}
+
```
Comment (akoeplinger) [inline mono/utils/mono-proclib.c]: The changes here will need to be done to proflog.c too to keep them in sync.

**Comment by akoeplinger** [comment]
@ludovic-henry this is still failing because HOST_ARM isn't defined in config.h, only TARGET_ARM.

**Comment by ludovic-henry** [comment]
@akoeplinger I fixed the build to define HOST_ARM properly

**Comment by akoeplinger** [comment]
@ludovic-henry thanks, LGTM now!

**Comment by lewurm** [comment]
this breaks `monodroid`:
```
[1:21:12] /Users/builder/data/lanes/1845/4e86d277/source/mono/mono/utils/mono-proclib.c: In function 'mono_cpu_count':
[1:21:12] /Users/builder/data/lanes/1845/4e86d277/source/mono/mono/utils/mono-proclib.c:638:3: error: 'count' undeclared (first use in this function)
[1:21:12]    count = strtol (((char*)buffer) + 2, NULL, 10);
[1:21:12]    ^
[1:21:12] /Users/builder/data/lanes/1845/4e86d277/source/mono/mono/utils/mono-proclib.c:638:3: note: each undeclared identifier is reported only once for each function it appears in
[1:21:12] make[7]: *** [mono-proclib.lo] Error 1
```

**Comment by akoeplinger** [comment]
@lewurm Thanks, fixed in 1180491691b59403122d656b7d6df6fe768f6846

**Comment by lewurm** [comment]
:+1:

**Pull request merged**